enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+
+None that I can think of at the moment
+
+### Test Plan
+
+- Add unit tests to the Cluster Ingress Operator to make sure the
+IngressController resource is generated as exepected.
+
+- Add periodic nightly tests which install single-node in the cloud, add a few
+worker nodes to it, then run conformance tests to make sure we don't run into
+any problems not described in this enhancement.
+
+- Add periodic nightly tests which install a single-node "none"-platform cluster,
+add worker nodes to it, and check that ingress traffic still works as expected
+and recovers even after the `router-default` pod gets deleted and rescheduled.
+Make sure this is still true even after upgrades.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after the single control-plane node reboots by
+running conformance tests post-reboot.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after an upgrade by running conformance tests
+post-upgrade.
+
+TODO: Describe day-1-workers tests?
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+TBD
+
+#### Dev Preview -> Tech Preview
+
+TBD
+
+#### Tech Preview -> GA
+
+TBD
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+In the non-goals section it's mentioned that this enhancement does not apply to
+clusters which have been installed prior to the enhancement, so their upgrade
+is not discussed.
+
+This enhancement, to the best of my knowledge, should have no problems
+persisting across cluster upgrades. The Test Plan section describes how this
+will be tested.
+
+### Version Skew Strategy
+
+Does not apply, to the best of my understanding.
+
+### Operational Aspects of API Extensions
+
+- The `ControlPlane` value of the `IngressPlacement` parameter value may not be
+used when the cluster's `ControlPlaneTopology` parameter is set to `External`,
+this combination is currently considered invalid. This may change in the future.
+
+- This API change only affects the defaulting behavior of the `IngressController` CR,
+it does not add any new capabilities to OCP, or give any more flexibility than there
+already was.
+
+- For administrators and support engineers, the `IngressController` is still the
+source of truth and where you need to look if you seek to understand the router
+placement in practice. Nothing has changed in that regard.
+
+#### Failure Modes
+
+- The `ControlPlane` value of the  `IngressPlacement` parameter value may not be
+used if the cluster's `ControlPlaneTopology` parameter is set to `External`.
+The Ingress Operator would treat such combination as if the `IngressPlacement`
+value is actually `Workers` instead, and log a warning to indicate that this
+invalid combination has been detected. The Ingress Operator should not fail
+to reconcile `IngressController` CRs due to this invalid combination.
```

Comment:
This seems like we are hiding an issue with the configuration. I would expect in this case that we would surface the issue, not do any further reconciliation and make the user fix the invalid configuration rather than hiding it and just carrying on with what we think is the right behaviour.

In general, if a user specifies an invalid configuration, we must surface that to them, else they might believe they've set a correct configuration and believe we are doing something that we are in fact not doing

Reply:
Discussed previously here https://github.com/openshift/enhancements/pull/1041#r817908736